### ISSUE TITLE
What is TCP socket bind failed. Rescheduling in log?

### ISSUE THREAD

**Issue by igroost** [acassen/keepalived#1488]
We started using keepalived on centos 7 and in the logs there was a message on a TCP socket bind failed. Rescheduling. What does it mean?

**Comment by pqarmitage** [comment]
If there isn't a relevant message immediately before the "TCP socket bind failed" massage, it means that the connect() call in socket_bind_connect() in keepalived/core/layer4.c failed with a reason other than EINPROGRESS, ENETUNREACH, EHOSTUNREACH, ECONNREFUSED, EHOSTDOWN, ENETDOWN, ECONNRESET, ECONNABORTED, ETIMEDOUT.

The log message is produced from tcp_connect_thread() in keepalived/check/check_tcp.c

**Comment by igroost** [comment]
@pqarmitage As far as I understand this message can be understood as verification failed?

**Comment by pqarmitage** [comment]
@igroost If the remote system was not reachable, or the service was not running on it, we would expect one of the errors ENETUNREACH, EHOSTUNREACH, ECONNREFUSED, EHOSTDOWN, ENETDOWN, ECONNRESET, ECONNABORTED, ETIMEDOUT. You are getting a different error, but unfortunately keepalived doesn't log what the error is.

I have pushed commit 4e33f7e to https://github.com/pqarmitage/keepalived/tree/updates but not yet merged it. This commit will log what the error is, so if you want to pull that and build keepalived with that commit, then we should be able to see what the error is.

**Comment by igroost** [comment]
@pqarmitage Thanks for the explanation.

If I understand the logic of keepalived correctly. This record is generated when you run health check. During connect execution, some error occurs that is different from the listed ones. What verification status is returned to keepalived? As far as I understand, it is returned that the check failed?

**Comment by pqarmitage** [comment]
@igroost When the connect() call fails with an error other than one of the errors listed above, it is treated as neither failure nor success, but as though the checker has not run, and the state of the checker will not change. So, if the checker fails from the start and alpha mode is in use, the real server will never be enabled; if alpha mode is not in use the real server will not be removed.

I think we need to know what error is being returned by connect() so we can work out how to deal with it. I have now merged commit 4e33f7e, and it would be really helpful if you could build and run the latest version from the master branch so we can find out what the error is that you are experiencing.

**Comment by igroost** [comment]
@pqarmitage I collected and launched keepalived from git yesterday, but so far only one socket error message has appeared in the log, 
but as I understand it, this is the expected behavior in the code.

Jan 30 10:02:54 ipvs-lan-new Keepalived_healthcheckers[31765]: DNS_CHECK ([10.10.109.51]:udp:53) failed to read socket. Connection refused

In order to see messages about checks in the log, I needed to enable the notification option for mail and checker_log_all_failures without this, only messages 
about the VIP and RS configuration were in the log

The attachment contains the keepalived log file where the version and settings are visible. Below in the log are messages about checks, they are mostly correct so I removed them.

We use keepalived version 2.0.0.  How correctly to compare the behavior of our and the latest version of keepalived. I use the same config file in both cases.

Can you tell me if it is possible to track what is happening during the connect call on our main keepalived, for example via strace? 
I tried, but I didn't get to see this exact moment.
[keepalived_log.txt](https://github.com/acassen/keepalived/files/4133055/keepalived_log.txt)

**Comment by pqarmitage** [comment]
@igroost Is 10.10.109.51 a local address on the system where keepalived is running? I think that is the only way it is possible to get Connection refused. This would mean that nothing is listening on UDP port 53 on the system, i.e. no named server is running.

You shouldn't need to turn on the notification option for mail and checker_log_all_failures to see the extra log information.

I have added commit b253df5 that refines the logging of connection errors. If you can build including that commit, you will need to add the `-D` command line option to increase the logging level. You should also include commit bbda1d8.

Then we will need to see is another of the `TCP socket bind failed. Rescheduling.` errors that was what first caused this issue to be opened. We should then also have the errno returned from the connect() call that will tell us more about what was happening.

**Comment by igroost** [comment]
@pqarmitage I collected keepalived today with the changes you added, I think on Monday I will report whether it was possible to catch the situation in which the message occurs.

> Is 10.10.109.51 a local address on the system where keepalived is running? I think that is the only way it is possible to get Connection refused. This would mean that nothing is listening on UDP port 53 on the system, i.e. no named server is running.

Yes, the address is 10.10.109.51 local and the service works on port 53.

**Comment by pqarmitage** [comment]
The `DNS_CHECK ([10.10.109.51]:udp:53) failed to read socket. Connection refused` message was probably caused by no DNS server running on the system at the time (or else the packet for some reason could not be delivered to the DNS server). keepalived is behaving as it should in this respect.

In respect of the `TCP socket bind failed. Rescheduling.` message, even if you don't get it again, at least the code has been improved to log the error in case anyone experiences it in the future.

**Comment by igroost** [comment]
@pqarmitage thanks for information.

Can you tell me if it is possible to track what is happening during the connect call on our main keepalived, for example via strace?

**Comment by pqarmitage** [comment]
@igroost You could use strace to see what return code is being returned from connect() and any associated errno, but with commit b253df5 keepalived should be reporting the errno anyway. You could also try using tcpdump or wireshark to trace the packets if that is the level you want to look at it, or you could use the iptables TRACE target, but that is a bit harder.

**Comment by igroost** [comment]
@pqarmitage thanks for information.

**Comment by igroost** [comment]
@pqarmitage We are still testing the latest keepalived build in our infrastructure.

In parallel, I tried to find a correlation of TCP socket bind failed messages. Rescheduling. 
In strace, I filtered only connect calls and got something like this output.

```
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("10.2.19.4")}, 128) = -1 EINPROGRESS (Operation now in progress)
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(443), sin_addr=inet_addr("10.9.5.51")}, 128) = -1 EINPROGRESS (Operation now in progress)
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("10.10.109.50")}, 128) = 0
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("10.1.23.31")}, 128) = -1 EINPROGRESS (Operation now in progress)
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("10.2.24.2")}, 128) = -1 EINPROGRESS (Operation now in progress)
16:03:43 connect(142, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("10.10.36.60")}, 128) = -1 EINPROGRESS (Operation now in progress)
```


As I understand it, calls with the status EINPROGRESS are processed by the code, and a call to the 
address 10.10.109.50 can just lead to a similar message?

**Comment by igroost** [comment]
@pqarmitage I was able to understand exactly why there are socket messages. This is affected by the VIP IPv6 
configuration. the service itself is probably no longer active on 
the real server. In strace, the call looks like this.

`16:50:52 connect(185, {sa_family=AF_INET6, sin6_port=htons(443), inet_pton(AF_INET6, "2001:67c:25d0:242::aa2", &sin6_addr), sin6_  owinfo=0, sin6_scope_id=0}, 128) = -1 ENETUNREACH (Network is unreachable)`

this information may be useful to you

**Comment by pqarmitage** [comment]
The connect() call to 10.10.109.50 is for UDP port 53, and I don't think EINPROGRESS will ever be returned for UDP.

The ENETUNREACH from the connect to 2001:67c:25d0:242::aa2 is also handled by the code, and it means that the system has no route to 2001:67c:25d0:24::/64 (assuming a /64 netmask), and presumably also no default route.

**Comment by igroost** [comment]
@pqarmitage Yes, on a machine with keepalived there is really no IPv6 connectivity because we refused to use IPv6 VIP, and forgot to remove it from the configuration. After I removed the IPv6 configuration, the TCP socket bind failed. Rescheduling. messages disappeared. 

As I understand it, because after running the keepalived check, the ENETUNREACH status returned because it is processed by the code, a more meaningful message should come to the log, and not TCP socket bind failed. Rescheduling. Correct me if I misunderstood.

**Comment by pqarmitage** [comment]
@igroost Many apologies, when commenting above on how the code works, I was describing the behaviour of the current code, not of v2.0.0 which I now see you are using.

**CURRENT BEHAVIOUR OF CODE**
I don't think your understanding is quite correct. ENETUNREACH is one of the errors that does NOT result in "TCP socket bind failed". ENETUNREACH will result in the check being considered to have failed. "TCP socket bind failed" is logged when connect() returns an unexpected error that is not explicitly handled by the code, and the check is considered to have neither passed or failed, but will be run again (i.e. the rescheduled).

**BEHAVIOUR OF v2.0.0**
Any error other than EINPROGRESS reported "TCP socket bind failed".

Commit 5860cf3 changed the way that ENETUNREACH is handled, so that it no longer reports "TCP socket bind failed", but treats it as the checker failed. This is my removing your IPv6 configuration stops "TCP socket bind failed".

I strongly recommend that you upgrade to v2.0.20. It is a very stable version and has many bug fixes and feature enhancements compared to v2.0.0. If you read the [Changelog](https://github.com/acassen/keepalived/blob/master/ChangeLog) it will show you all the changes by version.

**Comment by igroost** [comment]
@pqarmitage Thank you very much for the detailed explanation. Now I understand why we see this behavior. We plan to update to version 2.0.20. I think the ticket can be closed.
